### ISSUE TITLE
Does not work with current serilog version (4.1.0)

### ISSUE THREAD

**Issue by AlaskanDruid** [Jandini/serilog-enrichers-classname#3]
Always spits out blank for [{ClassName}]

**Comment by Jandini** [comment]
@AlaskanDruid I will look into that tonight. Thank you for reporting.

**Comment by Jandini** [comment]
@AlaskanDruid Can you provide an example ? 

I've tried with following code: 

`HelloWorld.csproj`
```xml
<Project Sdk="Microsoft.NET.Sdk">
	<PropertyGroup>
		<OutputType>Exe</OutputType>
		<TargetFramework>net8.0</TargetFramework>
		<BaseOutputPath>..\..\bin</BaseOutputPath>
		<ImplicitUsings>enable</ImplicitUsings>
		<PublishSingleFile>true</PublishSingleFile>
		<SelfContained>true</SelfContained>		
		<RuntimeIdentifier>win-x64</RuntimeIdentifier>
		<SatelliteResourceLanguages>en</SatelliteResourceLanguages>
	</PropertyGroup>
	<PropertyGroup Condition=" '$(Configuration)' == 'Release' ">
		<DebugType>none</DebugType>
		<DebugSymbols>false</DebugSymbols>
	</PropertyGroup>
	<ItemGroup>
		<PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="8.0.1" />
		<PackageReference Include="Microsoft.Extensions.Logging" Version="8.0.1" />
		<PackageReference Include="Serilog" Version="4.0.1" />
		<PackageReference Include="Serilog.Enrichers.ClassName" Version="0.2.0" />
		<PackageReference Include="Serilog.Extensions.Logging" Version="8.0.0" />
		<PackageReference Include="Serilog.Sinks.Console" Version="6.0.0" />
		<PackageReference Include="Serilog.Enrichers.Environment" Version="3.0.1" />
	</ItemGroup>
</Project>
```

`Main.cs`
```c#
using Microsoft.Extensions.Logging;

namespace HelloWorld.Program;

internal class Main
{
    private readonly ILogger<Main> _logger;

    public Main(ILogger<Main> logger)
    {
        _logger = logger;
    }

    public async Task RunAsync(CancellationToken cancellationToken)
    {
        _logger.LogInformation("Hello, World!");
        await Task.Delay(1000, cancellationToken); 
    }
}
```

`Program.cs`
```c#
// Created with JandaBox 0.8.1 http://github.com/Jandini/JandaBox
using HelloWorld.Program;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using Serilog;
using Serilog.Enrichers.ClassName;
using Serilog.Sinks.SystemConsole.Themes;

using var serviceProvider = new ServiceCollection()
    .AddLogging(builder => builder.AddSerilog(new LoggerConfiguration()
        .Enrich.WithMachineName()
        .Enrich.WithClassName()
        .WriteTo.Console(
            theme: AnsiConsoleTheme.Code,
            outputTemplate: "{Timestamp:yyyy-MM-dd HH:mm:ss.fff zzz} [{Level:u4}] [{SourceContext}] [{MachineName}] [{ClassName}] {Message}{NewLine}{Exception}")
        .CreateLogger()))
    .AddTransient<Main>()
    .BuildServiceProvider();

try
{
    using var cancellationTokenSource = new CancellationTokenSource();

    Console.CancelKeyPress += (sender, eventArgs) =>
    {
        serviceProvider.GetRequiredService<ILogger<Program>>()
            .LogWarning("User break (Ctrl+C) detected. Shutting down gracefully...");
        
        cancellationTokenSource.Cancel();
        eventArgs.Cancel = true; 
    };

    await serviceProvider.GetRequiredService<Main>().RunAsync(cancellationTokenSource.Token);
}
catch (Exception ex)
{
    serviceProvider.GetService<ILogger<Program>>()?
        .LogCritical(ex, "Program failed.");
}
```

**Comment by Jandini** [comment]
@AlaskanDruid I've discovered that top level statement does not generate SourceContext property. 
I'm not sure if that is your case until you provide the example. 

Meanwhile I'm preparing simple solution which will add SourceContext based on the caller: 

```
 public static LoggerConfiguration WithClassName(this LoggerEnrichmentConfiguration enrich, [CallerFilePath] string caller = "")
 {
     _ = enrich ?? throw new ArgumentNullException(nameof(enrich));

     enrich.WithProperty("SourceContext", Path.GetFileNameWithoutExtension(caller));                
     return enrich.With<ClassNameEnricher>();
 }
```

Please let me know if that is your case...

**Comment by AlaskanDruid** [comment]
Good Morning!

I apologize for the long time since my submission. I should have replied sooner :( I discovered the SourceContext option and went that direction after posting. However, it has to be manually set and it can get a bit wordy, which I don't like at all. 

... ... I wrote out a huge monster post with examples and everything as I was using the AppSettings.json file for configuration (I hate having hard coded config settings).... when... I could just do what you did above and hard code it heh. Same results though.. this is a Console app with "Do not use top-level statements" checked.

My apologies for all the images. code and quote tags are not working here for some reason :(

![image](https://github.com/user-attachments/assets/af4a1b56-ccd3-4cf1-a344-76b8dc4ec4af)

So this works IF I set the ForContext... but it does not work otherwise... I would have thought it'll work automatically? Did I assume wrong?

![image](https://github.com/user-attachments/assets/52f6a384-5003-434b-9567-ba94a0451090)

**Comment by Jandini** [comment]
@AlaskanDruid  So to clarify, this is not an issue of Serilog version? I will rename the issue to Enricher does not support top level statements.

**Comment by Jandini** [comment]
Since the default parameter might be confusing I propose following solution. 

```c#
public static LoggerConfiguration WithClassName(this LoggerEnrichmentConfiguration enrich)
{
    _ = enrich ?? throw new ArgumentNullException(nameof(enrich));

    var stackFrame = new StackTrace().GetFrame(1);
    var declaringType = stackFrame?.GetMethod()?.DeclaringType;

    enrich.WithProperty("SourceContext", declaringType?.FullName ?? "Unknown");
    return enrich.With<ClassNameEnricher>();
}
```

This will work the same as the previous solution and also support following case: 

```c#
using Serilog;
using Serilog.Enrichers.ClassName;
using SerilogDemo;

namespace Demo;

class MyProgram
{
    public static void Main(string[] args)
    {
        Log.Logger = new LoggerConfiguration()
            .MinimumLevel.Debug()
            .Enrich.WithClassName()
            .WriteTo.Console(outputTemplate: "[{Timestamp:HH:mm:ss} {Level:u3}] [{ClassName}] [{SourceContext}] {Message:lj}{NewLine}{Exception}")
            .CreateLogger();

        Log.Information("Application starting...");

        // Log from another class
        var demoClass = new DemoClass();
        demoClass.DoWork();

        Log.Information("Application ending...");
        Log.CloseAndFlush();
    }
}
```

Are you happy with it ? 

```
[15:22:46 INF] [MyProgram] [Demo.MyProgram] Application starting...
[15:22:47 INF] [DemoClass] [SerilogDemo.DemoClass] Doing some work...
[15:22:47 INF] [MyProgram] [Demo.MyProgram] Application ending...
```

**Comment by AlaskanDruid** [comment]
Ah sorry, I'm more at the level of using libraries and tools and not coding them :( I'm good with whatever you choose, really. Either way looks like there is no need for the line "logger.ForContext<T>" any more. Which is great!

**Comment by Jandini** [comment]
It's published... https://github.com/Jandini/serilog-enrichers-classname/releases/tag/0.3.0 

https://www.nuget.org/packages/Serilog.Enrichers.ClassName

@AlaskanDruid if you're happy with it.

btw, I've reduced one move line :) #5

**Comment by AlaskanDruid** [comment]
Thank you! I have updated. It looks like for the main class "Program", this works great. But any other class in the project, I need to use var Logger = Log.ForContext<(class name)>.  That is ok. I'm reverting back to using a logger instance per class with ForContext per class.  TY for the try though!

**Comment by Jandini** [comment]
If you can send me the full code, so I can understand the problem. The confusing bit is "I need to use var Logger = Log.ForContext<(class name)>. ".

**Comment by AlaskanDruid** [comment]
Ah yes! That would be useful. I should have thought of that :( Here they are below:

Project: 
![image](https://github.com/user-attachments/assets/cd981b6a-21bd-4a3e-a025-66704b11908b)

Class1.cs:

![image](https://github.com/user-attachments/assets/a800c6a3-7b67-4ddb-8ef7-0dcb44ec217d)

Class2.cs:

![image](https://github.com/user-attachments/assets/7e8dd1b4-59b0-4405-8aa6-907d0b90d317)


Program.cs

![image](https://github.com/user-attachments/assets/7a4cc00b-4917-48fe-b6e7-36c5eaf63e65)


Console Results:

![image](https://github.com/user-attachments/assets/001be2f2-5b78-44a1-8bd5-8e7ae9b381d9)

**Comment by Jandini** [comment]
It seems you are expecting something form Serilog that is not there. You are using a static "global" logger  in Serilog instead of injecting logger into your classes. Look at your SourceContext in your screenshoot. The class name is built of SourceContext. 

Your classes should have Logger injected into it.  

```C#
public class Class1
{
    private static readonly ILogger _logger = Serilog.Log.ForContext<Class1>();

    public void DoWork()
    {
        _logger.Information("Class1 is doing work.");
    }
}
```

Why don't you use dependency injection ? 

Try this...  https://github.com/Jandini/JandaBox

See how simple it is... you can write your own but here is something out of the box... 
```
dotnet new install JandaBox
dotnet new consolebox -n HelloWorld -b
```

**Comment by AlaskanDruid** [comment]
Hmm okay. I was trying to remove the need of ForContext<T> in order to spit out the class name as that is currently built into Serilog. I suppose the good news is that when using ForContext<T> with this library, the ClassName variable is filled with the ClassName (without the class path), which is great!
